rubix-bookkeeper/src/test/java/com/qubole/rubix/bookkeeper/TestWorkerBookKeeper.java

```diff
@@ -0,0 +1,189 @@
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class TestWorkerBookKeeper
+{
+  private static final Log log = LogFactory.getLog(TestBookKeeperServer.class.getName());
+  private static final String cacheTestDirPrefix = System.getProperty("java.io.tmpdir") + "/workerBookKeeperTest/";
+
+  private final Configuration conf = new Configuration();
+
+  @BeforeClass
+  public void initializeCacheDirectories() throws IOException
+  {
+    CacheConfig.setCacheDataDirPrefix(conf, cacheTestDirPrefix);
+
+    Files.createDirectories(Paths.get(cacheTestDirPrefix));
+    for (int i = 0; i < CacheConfig.getCacheMaxDisks(conf); i++) {
+      Files.createDirectories(Paths.get(cacheTestDirPrefix, String.valueOf(i)));
+    }
+  }
+
+  /**
+   * Verify that WorkerBookKeeper throws the correct exception when asked to handle heartbeats.
+   *
+   * @throws FileNotFoundException if the parent directory for the cache cannot be found when initializing the BookKeeper.
+   * @throws InterruptedException if the current thread is interrupted while starting the BookKeeper server.
+   */
+  @Test(expectedExceptions = UnsupportedOperationException.class)
+  public void testHandleHeartbeat() throws FileNotFoundException, InterruptedException
+  {
+    CacheConfig.setServiceRetryInterval(conf, 2000);
+    CacheConfig.setOnMaster(conf, true);
+
+    startBookKeeperServer();
+    final WorkerBookKeeper workerBookKeeper = new WorkerBookKeeper(conf, new MetricRegistry());
+    workerBookKeeper.handleHeartbeat("");
+  }
+
+  /**
+   * Verify that the heartbeat service correctly makes a connection using a BookKeeper client.
+   *
+   * @throws TTransportException if the BookKeeper client cannot be created.
+   * @throws InterruptedException if the current thread is interrupted while starting the BookKeeper server.
+   */
+  @Test
+  public void testHeartbeatRetryLogic_noRetriesNeeded() throws TTransportException, InterruptedException
+  {
+    final BookKeeperFactory bookKeeperFactory = mock(BookKeeperFactory.class);
+    CacheConfig.setServiceRetryInterval(conf, 2000);
+    CacheConfig.setOnMaster(conf, true);
+
+    when(bookKeeperFactory.createBookKeeperClient(anyString(), ArgumentMatchers.<Configuration>any())).thenReturn(
+        new RetryingBookkeeperClient(
+            new TSocket("localhost", CacheConfig.getServerPort(conf), CacheConfig.getClientTimeout(conf)),
+            CacheConfig.getMaxRetries(conf)));
+
+    startBookKeeperServer();
+    final WorkerBookKeeper.HeartbeatService heartbeatService = new WorkerBookKeeper.HeartbeatService(conf, bookKeeperFactory);
+  }
+
+  /**
+   * Verify that the heartbeat service correctly makes a connection using a BookKeeper client after a number of retries.
+   */
+  @Test
+  public void testHeartbeatRetryLogic_connectAfterRetries()
+  {
+    final int retryInterval = 500;
+    CacheConfig.setServiceRetryInterval(conf, retryInterval);
```

Comment:
as BeforeMethod has already started the server, lets stop it first and then invoke startBookKeeperServerWithDelay